# Binomial Tree Example

In [12]:
import numpy as np
from numpy import exp

In [47]:
u_factor = 1.20
d_factor = 0.80
r = 0.00
K = 110
S0 = 100

q = (exp(r) - d_factor) / (u_factor - d_factor)

In [48]:
S3 = np.array([172.8, 115.2, 76.8, 51.2])
payoffs = np.maximum(S3 - K, 0); payoffs

array([ 62.8,   5.2,   0. ,   0. ])

In [49]:
ExpQ = lambda payoffs, q, r: exp(-r) * np.sum(payoffs * np.array([q, 1 - q]))

## Payoffs

### $t = 3$ 

In [93]:
fuuu, fuud, fudd, fddd = payoffs

### $t=2$

In [50]:
# Node 3, at t=2
fuu = ExpQ((62.8, 5.2), q, r)

# Node 2, at t=2
fud = ExpQ((5.2, 0), q, r)

# Node 1, at t=2
fdd = ExpQ((0, 0), q, r)

### $t=1$

In [51]:
# Node 2, at t=1
fu = ExpQ((fuu, fud), q, r)

# Node 1, at t=1
fd = ExpQ((fud, fud), q, r)

### $t=0$

In [52]:
f = ExpQ((fu, fd), q, r); f

10.450000000000001

## Hedging the derivative

$$
    \phi = \frac{f_u - f_d}{S(u -d)}
$$

In [82]:
# Hedging at t = 0
phi1 = (fu - fd) / (S0 * (u_factor - d_factor)) # number of shares to buy
psi1 = phi1 * S0 - f # amount to borrow

phi1, psi1

(0.39250000000000007, 28.800000000000004)

We need $0.3925$ units of $S$ and borrow $28.8$ to cover the expenses

In [92]:
# Suppose it went up
phi2 = (fuu - fud) / (S0 * u_factor * (u_factor - d_factor)) 
psi2 = (phi2 - phi1) * (S0 * u_factor) + psi1

phi2, psi2

(0.65416666666666679, 60.20000000000001)

We now need around $0.65$ units of $S$ and it costs an additional $31.4$ which brings the total expenses up to $60.2$

In [95]:
# Suppose it went up one more time
phi3 = (fuuu - fuud) / (S0 * u_factor ** 2 * (u_factor - d_factor))
psi3 = (phi3 - phi2) * (S0 * u_factor ** 2) + psi2

phi3, psi3

(1.0000000000000004, 110.00000000000006)

At $t=2$, we need to have one unit of $S$ and the total cost of the position is $110$

## $t=3$: The stock falls
Suppose, at $t=3$, that the stock falls

In [99]:
# The total value of the position in the stock is
(100 * u_factor ** 2 * d_factor) * phi3

115.20000000000006

In [101]:
# We have borrowed a total amount of
- psi3

-110.00000000000006

In [102]:
# Which leads to a total portfolio value of
(100 * u_factor ** 2 * d_factor) * phi3 - psi3

5.2000000000000028

The latter being the value of the derivative at $t=3$. We can thus fulfill our obligation. Had the stock gone up was last time, the value of th portfolio would've been the necesary to also fulfill the obligation.

In [96]:
fuuu, fuud

(62.800000000000011, 5.2000000000000028)